### DISCUSSION TITLE
Statistics

### DISCUSSION THREAD

**Discussion by mimmiechan** [espanso/espanso#1184] [Feature requests and Ideas]
It would be a nice feature if espanso could tell you how many times it has helped you i.e. the number of times espanso has been triggered and replaced something.

Now that I think about it, it would actually be useful and interesting if you could see which triggers you use the most. I could see some privacy issues with this though.

Another feature, that I don't think is likely to happen but could still be incredibly useful: that espanso could detect when it has done something the user doesn't want. I guess the most straightforward way to do this would be to listen for Ctrl+Zs pressed after an espanso replacement. But I don't think the user is likely to use Ctrl+Z everytime espanso "messed up", I think they are more likely to use the backspace button. But a backspace click might just indicate that user just wasn't happy with what he wrote, with or without espanso's help.
This data could be used to try and detect problematic triggers and perhaps even try to correct it, or give suggestions for correction. I use espanso for spell correction and for replacing some chat speech words and shorthands with the "correct" words, just trying to minimize the number of keystrokes required as much as possible. But sometimes I add something to my config that wasn't well-thought-out and I end up with false positives. So what if espanso would show a pop up-window in a case like that, like "you seem to often undo what this trigger does, do you want to modify it? [make it a Word Trigger] [propagate case] [edit] [remove]". Or maybe just show a notification that says you have some unsolved issues that you can access from the tray icon where you are get these prompts, but you go there manually instead of being nagged and required to take action when you are in the middle of something. Might be kind of feature creep. But statistics could prove useful for a number of things.

EDIT: and to add to my AI nonsense stuff: maybe it could keylog to detect phrases (or really long words) that the user often seem to write and give an option for the user to add those phrases to triggers. For example, imagine you do not have a trigger for "With regards, X", espanso would notice that you keep typing this out manually all the time over and over. Eventually it would just say "hey, are you sure you don't want to _espanso_ that?"

This issue started with a very simple feature request and escalated to basically a full-blown AI implementation project. sorry.

**Comment by federico-terzi** [comment]
Hey @mimmiepower,

Thanks for your feedback!

I've been thinking about statistics for a while (PhaseExpress does it) and I definitely like the idea.

Regarding the two other ideas, indeed an automatic suggestion system built-in into espanso would be incredibly beneficial, but also pretty complex to implement. Moreover, it poses some non-trivial security risks, as in order to track which phrases/sentences are typed the most, we basically have to implement a keylogger.

I'm not closing this issue as won't fix because I think some of your ideas might become part of espanso in the future, but definitely not in the short term :)

Cheers :)

**Comment by 4lch4** [comment]
I really like the idea of statistics as well, that'd be neat. There is obviously a lot that comes with that though as previously mentioned.

I'm not too familiar with the code for espanso so I'm speakin' out of my ass a bit here, could it be a trigger upon completing an expansion to help avoid privacy issues. Instead of espanso monitoring for stats, it's an event triggered when a completion is done. In my case, `:jbh` is a shortcut for J. B. Hunt, the company I work at. When it expands, it could add +1 to the `:jbh` expansion. Again, that's mostly me blowing smoke as I'm not familiar with the code yet 😅

**Comment by federico-terzi** [comment]
There are a lot of ways in which we can implement it. One little thing that I really like about PhaseExpress is that it shows a message in the context menu such as:

```
You saved 231 hours of typing with PhaseExpress
```

And even though it doesn't provide any actual benefit, it's cool anyway :)

**Comment by grenudi** [comment]
Same. I'm seeking a way to get usage statistics for individual snippets and triggers. It would benefit a lot in my job. 
Any simple solution would do.
@federico-terzi  , maybe it would be possible to implement something simple, accessible through command line interface ?  
Or I could add script extension into each snippet, though, it would be a little bit cumbersome.
